src/anomalib/metrics/spro.py

```diff
@@ -0,0 +1,158 @@
+"""Implementation of SPRO metric based on TorchMetrics."""
+
+# Copyright (C) 2022 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+import logging
+
+import torch
+from torchmetrics import Metric
+from torchmetrics.utilities.data import dim_zero_cat
+
+logger = logging.getLogger(__name__)
+
+
+class SPRO(Metric):
+    """Saturated Per-Region Overlap (SPRO) Score.
+
+    This metric computes the macro average of the saturated per-region overlap between the
+    predicted anomaly masks and the ground truth masks.
+
+    Args:
+        threshold (float): Threshold used to binarize the predictions.
+            Defaults to ``0.5``.
+        saturation_config (dict): Saturations configuration for each label (pixel value) as the keys.
+            Defaults: ``None`` (which the score is equivalent to PRO metric, but with the 'region' are
+            separated by mask files.
+        kwargs: Additional arguments to the TorchMetrics base class.
+
+    Example:
+        Import the metric from the package:
+
+        >>> import torch
+        >>> from anomalib.metrics import SPRO
+
+        Create random ``preds`` and ``labels`` tensors:
+
+        >>> labels = torch.randint(low=0, high=2, size=(1, 10, 5), dtype=torch.float32)
+        >>> preds = torch.rand_like(labels)
+
+        Compute the SPRO score for labels and preds:
+
+        >>> spro = SPRO(threshold=0.5)
+        >>> spro.update(preds, _, labels)
+        >>> spro.compute()
+        tensor(0.6333)
+
+        .. note::
+            Note that the example above shows random predictions and labels.
+            Therefore, the SPRO score above may not be reproducible.
+
+    """
+
+    targets: list[torch.Tensor]
+    preds: list[torch.Tensor]
+
+    def __init__(self, threshold: float = 0.5, saturation_config: dict | None = None, **kwargs) -> None:
+        super().__init__(**kwargs)
+        self.threshold = threshold
+        self.saturation_config = saturation_config
+        self.add_state("preds", default=[], dist_reduce_fx="cat")
+        self.add_state("targets", default=[], dist_reduce_fx="cat")
+
+    def update(self, predictions: torch.Tensor, _: torch.Tensor, masks: torch.Tensor) -> None:
```

Comment:
both metric-collection and torch metric has ` def update(self, *args, **kwargs) -> None` signature. Would it be possible to remove `_: torch.Tensor`?

Reply:
I think there might be two possible solutions (or more?) if we want to remove `_: torch.Tensor`, but we still want to handle the case of simultaneously evaluating the dataset using existing metrics (PRO, etc) and new metrics (SPRO and AUSPRO).

#### 1. Separate SPRO and AUSPRO into different metric collection
Currently, when we put the SPRO and PRO into the same pixel metric collection, it means we need to handle both case of `mask` and `masks`. so that's why we need to pass both mask and masks to the update function of the metric collection like this:
https://github.com/openvinotoolkit/anomalib/blob/a87cdf533c3a8b1c744d081c6db51edc54baa95c/src/anomalib/callbacks/metrics.py#L185-L193

However, if we can separate the metric collection, we can call update(anomaly_map, mask) and update(anomaly_map, masks) separately, so that we can remove the `_`.

#### 2. Calling update without using the `update` function of the metric collection
Instead of using `pixel_metric.update` (AnomalibMetricCollection), we iterate the metrics in the collection via `pixel_metric.items(copy_state=False)`, and call update method of the torch metric directly. Thus, each metric can call either update(anomaly_map, mask) or update(anomaly_map, masks). And after this update, we need to set  `_updated_called` attribute of the AnomalibMetricCollection to True, since it is needed for this part:

https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/callbacks/metrics.py#L198-L202

I am not sure which one is better, but the first approach seems requires lots of modification in the codes. The second approach is more simple, but i haven't fully investigate it, if it is really okay to call the torch metric update directly without using the update method of the metric collection.

Reply:
I have removed `_: torch.Tensor` in the latest implementation, by using the 2nd approach. I have tested it, it is working as expected. 

https://github.com/openvinotoolkit/anomalib/blob/3bbc7507ea29f537118728b579e72e04b998f78f/src/anomalib/callbacks/metrics.py#L205-L214